File: lib/contracts.ts
import { z } from 'zod';
export const categories=['everyday','dining','class','appointment','entertainment','accommodation'] as const;
export const requirementInput=z.object({text:z.string().trim().min(3).max(1000),hard:z.boolean(),share_allowed:z.boolean()}).strict();
export const arrangementInput=z.object({
 title:z.string().trim().min(3).max(160),category:z.enum(categories),venue_name:z.string().trim().min(2).max(160),
 venue_url:z.union([z.literal(''),z.url().max(2000).refine(s=>/^https?:\/\//.test(s))]),contact_email:z.union([z.literal(''),z.email().max(254)]),
 timezone:z.string().max(100).refine(zone=>{try{new Intl.DateTimeFormat('en',{timeZone:zone});return true;}catch{return false;}}),
 visit_date:z.union([z.literal(''),z.iso.date()]),requirements:z.array(requirementInput).min(1).max(30),outreach_allowed:z.boolean(),
}).strict().refine(v=>!v.outreach_allowed||Boolean(v.contact_email),{message:'Add the venue email before allowing outreach.'});
export type ArrangementInput=z.infer<typeof arrangementInput>;
export const editArrangementInput=arrangementInput.safeExtend({requirements:z.array(requirementInput.extend({id:z.uuid().optional()})).min(1).max(30)});
export type EditArrangementInput=z.infer<typeof editArrangementInput>;
export type Visit={content_version:number;id:string;title:string;category:string;venue_name:string;venue_url:string;contact_email:string;timezone:string;visit_date:string|null;state:string;version:number};
export type ResearchReport={content_version:number;requirements_snapshot:Snapshot["requirements"];id:string;case_id:string;job_id:string;case_version:number;inquiry_draft:string;created_at:string;report:{mode:'live'|'local_test';model_id:string;summary:string;venue_match:'matched'|'uncertain'|'mismatch';sources:{id:string;url:string;title:string;retrieved_at:string}[];findings:{requirement_id:string;status:'unknown'|'source_supports'|'source_reports_unavailable'|'conflicting';explanation:string;citations:{source_id:string;quote:string}[]}[]}};
export type CaseRevision={id:string;case_id:string;previous_version:number;result_version:number;created_at:string;snapshot:{case:Visit;requirements:Snapshot["requirements"];permission:{recipient:string;outreach_allowed:boolean}}};
export type Snapshot={revisions:CaseRevision[];profile:{display_name:string;needs:string[]}|null;cases:Visit[];requirements:{id:string;case_id:string;text:string;hard:boolean;share_allowed:boolean;state:string}[];permissions:{case_id:string;recipient:string;outreach_allowed:boolean}[];jobs:{id:string;case_id:string;state:string;last_error?:string|null}[];reports:ResearchReport[];events:{id:number;case_id:string;kind:string;created_at:string}[]};

File: agent/verra_agent/contracts.py
from typing import Literal
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field


class Contract(BaseModel):
    model_config = ConfigDict(extra="forbid")


class Requirement(Contract):
    id: UUID
    text: str = Field(min_length=3, max_length=1000)
    hard: bool
    share_allowed: bool


class Visit(Contract):
    id: UUID
    version: int = Field(ge=1)
    title: str
    venue_name: str
    venue_url: str
    visit_date: str | None
    timezone: str


class Permission(Contract):
    version: int
    recipient: str
    outreach_allowed: bool


class Context(Contract):
    job_id: UUID
    case: Visit
    requirements: list[Requirement] = Field(min_length=1, max_length=30)
    permissions: Permission


class Claim(Contract):
    job_id: UUID
    lease_token: UUID


class Citation(Contract):
    source_id: str = Field(min_length=1, max_length=32)
    quote: str = Field(min_length=8, max_length=800)


class Finding(Contract):
    requirement_id: UUID
    status: Literal["unknown", "source_supports", "source_reports_unavailable", "conflicting"]
    explanation: str = Field(min_length=1, max_length=1200)
    citations: list[Citation] = Field(max_length=5)


class Analysis(Contract):
    venue_match: Literal["matched", "uncertain", "mismatch"]
    summary: str = Field(min_length=1, max_length=1500)
    findings: list[Finding] = Field(min_length=1, max_length=30)


class Source(Contract):
    id: str
    url: str
    title: str
    retrieved_at: str
    text: str


def normalized(text: str) -> str:
    return " ".join(text.split())


def validate_analysis(analysis: Analysis, context: Context, sources: list[Source]) -> Analysis:
    """Reject fabricated IDs/quotes; downgrade uncertain venue identity in code."""
    expected = {r.id for r in context.requirements}
    actual = [f.requirement_id for f in analysis.findings]
    if len(actual) != len(expected) or set(actual) != expected:
        raise ValueError("Result must cover exactly this visit's requirements")
    known = {s.id: s for s in sources}
    for finding in analysis.findings:
        for citation in finding.citations:
            source = known.get(citation.source_id)
            if not source or normalized(citation.quote) not in normalized(source.text):
                raise ValueError("Citation does not match retrieved evidence")
        if finding.status != "unknown" and not finding.citations:
            raise ValueError("A supported finding needs a source quotation")
    if not sources or analysis.venue_match != "matched":
        reason = "The source could not be matched to this venue and visit. These needs remain unverified."
        if not sources:
            reason = "No usable venue page was retrieved. These needs remain unverified."
        return Analysis(venue_match="uncertain" if not sources else analysis.venue_match, summary=reason,
                        findings=[Finding(requirement_id=r.id, status="unknown", explanation=reason, citations=[])
                                  for r in context.requirements])
    return analysis

